src/vcf/fixer.cpp

```diff
@@ -210,10 +238,149 @@ namespace ebi
+                } else if (first_key_occurrence[subfields[0]] != subfields[1]) {
+                    fields_to_remove.insert(subfields[0]);
+                }
+            }
+
+            std::vector<std::string> fixed_column;
+            size_t num_removed_duplicates = 0;
+
+            for (auto & ordered_key : ordered_keys) {
+                if (fields_to_remove.find(ordered_key) == fields_to_remove.end()) {
+                    fixed_column.push_back(ordered_key + key_value_separator + first_key_occurrence[ordered_key]);
+                } else {
+                    num_removed_duplicates++;
+                }
+            }
+
+            if (fixed_column.size() == 0) {       // all the fields were removed
+                output << empty_value;
+            } else {
+                util::print_container(output, fixed_column, "", separator, "");
+            }
+
+            return num_removed_duplicates;
+        }
+
+        size_t Fixer::remove_duplicate_format_sample_pairs(const std::string &string_line) {
+            const size_t format_column_index = 8;
+            size_t num_removed_duplicates = 0;
+
+            using iter = std::vector<std::string>::iterator;
+            fix_columns(format_column_index, -1, string_line, "\t", [&](iter first, iter last) {
+                std::map<std::string, std::vector<size_t>> format_fields_indexes;
+                std::vector<std::string> ordered_fields;
+                std::vector<std::string> format_keys;
+                util::string_split(*first, ":", format_keys);
+
+                for (size_t i = 0; i < format_keys.size(); i++) {
+                    format_fields_indexes[format_keys[i]].push_back(i);
+                    if (format_fields_indexes[format_keys[i]].size() == 1) {
+                        ordered_fields.push_back(format_keys[i]);
+                    }
+                }
+
+                std::set<std::string> fields_to_remove = get_format_fields_to_remove(format_fields_indexes, first, last);
+
+                std::vector<std::string> fixed_format;
+                for (auto & ordered_field : ordered_fields) {
+                    if (fields_to_remove.find(ordered_field) == fields_to_remove.end()) {
+                        // keep first occurrence, discard the rest if present
+                        fixed_format.push_back(ordered_field);
+                    } else {
+                        num_removed_duplicates++;
+                    }
+                }
+
+                if (fixed_format.empty()) {
+                    throw std::runtime_error("Could not fix FORMAT duplicate fields: All fields had to be removed, but missing value is not permitted");
+                }
+                util::print_container(output, fixed_format, "", ":","");
+
+                for (auto it = first + 1; it != last; it++) {
+                    std::vector<std::string> sample_fields;
+                    util::string_split(*it, ":", sample_fields);
```

Comment:
Even though in the other method pre-splitting the subfields doesn't give any advantage, in this one the split is done twice, one inside `get_format_fields_to_remove` and another one in this loop.

Considering that errors of this type will be probably systematic across a whole file, and that the number of columns could be in the order of thousands, splitting all those strings twice could really slow down the process.

I suggest the following process:

1. Benchmark how the current code takes to process let's say a file with 10,000 lines and 100 samples. It can be a synthetic file.
2. Change the code to pre-split all the samples' subfields and keep them in a data structure. `get_format_fields_to_remove` would need to receive that data structure as argument.
3. Benchmark the new code against the same file.

Given our previous experience with string splits, I expect to see a difference of at least a 10-15%.